lib/util.c

```diff
@@ -2187,9 +2194,332 @@ EXPORTED void xsyslog_fn(int priority, const char *description,
     errno = saved_errno;
 }
 
-EXPORTED char *modseqtoa(modseq_t modseq)
+static const char *xsyslog_ev_escape_value(struct buf *);
+static void  xsyslog_ev_fmt_skip_va(const char *, va_list);
+static char *xsyslog_ev_fmt_va(const char *, va_list);
+static char  xsyslog_ev_guess_printf_escape(const char *);
+
+/* xsyslog_ev:
+ *   1. use xsyslog_ev_fmt to format a string for syslogging
+ *   2. Then syslog it with the specified priority
+ *
+ * If something goes wrong in formatting, the function syslogs
+ * an unhappy message and continues.
+ */
+EXPORTED void xsyslog_ev(int priority, const char *event_name, ...)
 {
-    struct buf buf = BUF_INITIALIZER;
-    buf_printf(&buf, MODSEQ_FMT, modseq);
-    return buf_release(&buf);
+    va_list ap;
+    va_start(ap, event_name);
+
+    char *s = xsyslog_ev_fmt_va(event_name, ap);
+    va_end(ap);
+
+    if (!s) {
+        syslog(priority, "COULD NOT FORMAT SYSLOG STRING FOR EVENT '%s'", event_name);
+        return;
+    }
+
+    syslog(priority, "%s", s);
+    free(s);
+}
+
+/* xsyslog_ev_fmt:
+ *   1. Format a key-value list for syslogging
+ *   2. Allocate a string to hold the result
+ *   3. The CALLER is responsible for freeing this string
+ *   4. Returns NULL if something goes wrong
+ *
+ * EXAMPLE:
+ *
+ *     xsyslog_ev_fmt("earthquake",
+ *                    "location", "%s", "Dubuque",
+ *                    "richter", "%5.1f", 7.9)
```

Comment:
This doesn't document the final necessary NULL, we should in all of the examples, etc...